python/src/fast.cpp

```diff
@@ -183,4 +188,102 @@ void init_fast(nb::module_& parent_module) {
         Returns:
           array: The quantized version of ``w``
       )pbdoc");
+
+  m.def(
+      "metal_kernel",
+      [](std::string name,
+         std::string source,
+         std::map<std::string, nb::handle> inputs_,
+         std::map<std::string, std::vector<int>> output_shapes,
+         std::map<std::string, Dtype> output_dtypes,
+         std::tuple<int, int, int> grid,
+         std::tuple<int, int, int> threadgroup,
+         std::optional<std::map<std::string, nb::handle>> template_args_,
+         bool ensure_row_contiguous,
+         bool verbose,
+         StreamOrDevice s) -> std::map<std::string, array> {
+        std::map<std::string, fast::TemplateArg> template_args;
+        if (template_args_) {
+          for (auto [name, value] : template_args_.value()) {
+            // Handle bool, int and dtype template args
+            if (PyBool_Check(value.ptr())) {
+              bool bool_val = nb::cast<bool>(value);
+              template_args.insert({name, bool_val});
+            } else if (PyLong_Check(value.ptr())) {
+              int int_val = nb::cast<int>(value);
+              template_args.insert({name, int_val});
+            } else if (
+                nb::type_check(value.type()) &&
+                nb::type_info(value.type()) == typeid(Dtype)) {
+              Dtype dtype = nb::cast<Dtype>(value);
+              template_args.insert({name, dtype});
+            } else {
+              throw std::invalid_argument(
+                  "Invalid template argument. Must be `mlx.core.Dtype`, `int` or `bool`.");
+            }
+          }
+        }
+        std::map<std::string, array> inputs;
+        for (auto [name, in] : inputs_) {
+          if (name != "stream") {
+            auto value = nb::cast<ArrayInitType>(in);
+            auto arr = create_array(value, std::nullopt);
+            inputs.insert({name, arr});
+          }
+        }
+        return fast::metal_kernel(
+            name,
+            source,
+            inputs,
+            output_shapes,
+            output_dtypes,
+            grid,
+            threadgroup,
+            template_args,
+            ensure_row_contiguous,
+            verbose,
+            s);
+      },
+      "name"_a,
+      "source"_a,
+      "inputs"_a,
+      "output_shapes"_a,
+      "output_dtypes"_a,
+      "grid"_a,
+      "threadgroup"_a,
+      "template"_a = nb::none(),
+      "ensure_row_contiguous"_a = true,
+      "verbose"_a = false,
+      nb::kw_only(),
```

Comment:
Wdyt about making more of these arguments `kw_only`? Maybe even all of them?

Reply:
Missed this one -- I agree let's make them all kw_only.

Reply:
Yea.. that will also make the API more flexible since we won't need to worry about back-compatibility with positions in the argument list.